### ISSUE TITLE
Shader panic on Intel GPU

### ISSUE THREAD

**Issue by Connicpu** [glium/glium#635]
I've got two relatively simple shaders which work just fine on my desktop (nvidia), but cause a panic when I try to create a program. The hardware I'm running on is a Surface Pro 3. i7-4650U, which has an HD5000 gpu.

I have verified that both opengl and glsl supported versions are `Version(Gl, 4, 3)`, and it panics with the following message

```
thread '<main>' panicked at 'Debug message with high or medium severity: `Error has been generated. GL error GL_INVALID_ENUM`.
```

Here are the vertex and fragment shaders

``` glsl
#version 330

uniform mat4 camera;

in vec2 position;
in vec2 tex_coord;
in vec4 color;

out vec4 v_color;
out vec2 v_tex_coord;

void main() {
    gl_Position = vec4(position, 0.0, 1.0) * camera;
    v_tex_coord = tex_coord;
    v_color = color;
}
```

``` glsl
#version 330

uniform sampler2D texture;

in vec4 v_color;
in vec2 v_tex_coord;

void main() {
    gl_FragColor = v_color * texture2D(texture, v_tex_coord);
}
```

EDIT: The repository for the code I'm trying to run is [here](https://github.com/Connorcpu/rgame)

**Comment by tomaka** [comment]
Thanks for reporting!

Could you run your program with `RUST_BACKTRACE=1` in order to get a backtrace of when that happens?

**Comment by Connorcpu** [comment]
Yeah, it's just a windows 8.1 device. The backtraces don't seem quite right on windows though :/
```
~\s\r\rgame [master] § cargo run
     Running `target\rgame`
Initializing graphics
Window made!
Version(Gl, 4, 3)
Version(Gl, 4, 3)
Initializing sprite system
thread '<main>' panicked at 'Debug message with high or medium severity: `Error has been generated. GL error GL_INVALID_ENUM`.
Please report this error: https://github.com/tomaka/glium/issues', C:\Users\Connor\.cargo\git\checkouts\glium-cee1c79709e4d1bd\master\src\li
b.rs:1181
stack backtrace:
   1:           0x980fd4 - main
   2:           0x99b7c9 - main
   3:           0x8ed367 - main
   4:           0x8edb2b - main
   5:           0x65a152 - main
   6:           0x659d33 - main
   7:     0x7ffe0dfd75c6 - RegisterProcTableCallback
   8:     0x7ffe0dfd77ee - RegisterProcTableCallback
   9:     0x7ffe0dfd828a - RegisterProcTableCallback
  10:     0x7ffe0e12184e - RegisterProcTableCallback
  11:     0x7ffe0e117db4 - RegisterProcTableCallback
  12:     0x7ffe0e2eddc0 - DrvSetLayerPaletteEntries
  13:           0x52ae6f - main
  14:           0x52419b - main
  15:           0x5180d2 - main
  16:           0x516424 - main
  17:           0x42a563
  18:           0x4f5f14
  19:           0x99dc3c - main
  20:           0x99dc19 - main
  21:           0x99c0d5 - main
  22:           0x4f6084 - main
  23:           0x4013b5
  24:           0x4014e8
  25:     0x7ffe3cf516ad - BaseThreadInitThunk
An unknown error occurred
```

**Comment by tomaka** [comment]
Yeah, that's not very helpful.
Could you run your program with `apitrace trace <program>` and upload the .trace?

**Comment by Connorcpu** [comment]
I don't seem to be able to get apitrace to work, the program just immediately crashes with an access violation. Debugging shows it crashing somewhere relating to a critical section. (my program crashes, not apitrace)

**Comment by tomaka** [comment]
Is your version of glium old?
The reason why the backtrace is bad is maybe that it was because the change to glium that made it entirely single-threaded.

**Comment by Connorcpu** [comment]
I'm on the latest version of glium from git. The backtrace issue is a problem with how windows gets symbol information :/

**Comment by tomaka** [comment]
Does it also produce useless stack traces if you run with `gdb`? I'm on Windows too, and gdb works fine for me.

Could you also obtain the OpenGL version number by calling `display.get_gl_version()`?

**Comment by Connorcpu** [comment]
Both get_opengl_version and get_glsl_supported_version return Version(Gl, 4, 3).
All I could get from gdb is this. I haven't done much debugging with gdb before, is there something I can do with cargo to make it output more information?
```
Breakpoint 1, 0x00000000008ed950 in rt::unwind::begin_unwind_inner::hbfd1c4c7cd361ab07oJ ()
(gdb) bt full
#0  0x00000000008ed950 in rt::unwind::begin_unwind_inner::hbfd1c4c7cd361ab07oJ ()
No symbol table info available.
#1  0x00000000008ee4ab in rt::unwind::begin_unwind_fmt::h6f957c72afec51dfInJ ()
No symbol table info available.
#2  0x000000000023d748 in ?? ()
No symbol table info available.
#3  0x000000000023d748 in ?? ()
No symbol table info available.
#4  0x000000000065a000 in std..env..VarError::glue_drop.57707::h6a27e1a9c1954479 ()
No symbol table info available.
#5  0x000000000065a043 in Display::init_debug_callback::callback_wrapper::h0b4d845e1217dddbgPb ()
No symbol table info available.
#6  0x00007ffadb8b75c6 in ig75icd64!RegisterProcTableCallback () from C:\windows\SYSTEM32\ig75icd64.dll
No symbol table info available.
#7  0x0000000000000008 in ?? ()
No symbol table info available.
#8  0x000000000023dc90 in ?? ()
No symbol table info available.
#9  0x0000000000000000 in ?? ()
No symbol table info available.
```

**Comment by XMPPwocky** [comment]
Encountered something similar in vel0city on Intel GPU (Haswell). (Laptop has switchable graphics, so this isn't usually an issue.)

````
thread '<main>' panicked at 'Debug message with high or medium severity: `Error has been generated. GL error GL_INVALID_ENUM`.
Please report this error: https://github.com/tomaka/glium/issues', C:\Users\Nathaniel\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.2.1\src\context/mod.rs:516

Breakpoint 1, context::init_debug_callback::callback_wrapper::h0ca0bdd5e81351d0LDt ()
    at C:\Users\Nathaniel\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.2.1\src\context/mod.rs:519
519                 }
(gdb) bt
(gdb) Undefined command: "rbt".  Try "help".
bt
#0  context::init_debug_callback::callback_wrapper::h0ca0bdd5e81351d0LDt ()
    at C:\Users\Nathaniel\.cargo\registry\src\github.com-1ecc6299db9ec823\glium-0.2.1\src\context/mod.rs:519
#1  0x00007ffef1cf7736 in ig75icd64!RegisterProcTableCallback ()
   from C:\windows\SYSTEM32\ig75icd64.dll
#2  0x0000000000000008 in ?? ()
#3  0x00007fff17a01608 in ?? ()
#4  0x0000000000000000 in ?? ()
````

**Comment by kornholi** [comment]
I think this is the same issue as in #998 and was fixed by #1020
